library/game_state.py

```diff
@@ -0,0 +1,167 @@
+                f"Current: SubGrid {self.user_select_subgrid} "
+                f"| Space {self.user_select_space}"
+            )
+        elif self.user_input == "KEY_ENTER" and self.user_select_subgrid != 0:
+            self.term_info[1] = "Select Space by entering 1-9"
+            self.confirm_entry(term)
+            self.next = 30
+
+        self.redraw_user_term(term)
+
+    def update_space_select(self, term: blessed.Terminal, board: Board) -> None:
+        """Update the user selected space"""
+        self.current = 30
+
+        if self.user_input in ("1", "2", "3", "4", "5", "6", "7", "8", "9"):
+            self.user_select_space = int(self.user_input)
+            self.term_info[2] = (
+                f"Current: SubGrid {self.user_select_subgrid} "
+                f"| Space {self.user_select_space}"
+            )
+        elif self.user_input == "KEY_ENTER" and self.user_select_space != 0:
+            if self.confirm_good_move(board):
+                self.update_board = True  # good to place player token
+                self.confirm_entry(term)
+            else:
+                self.next = 30  # TODO go to error handling and reset values
+                self.term_info[
+                    1
+                ] = f"{term.red}*-That is an illegal move-*{term.normal}"
+                self.redraw_user_term(term)
+                time.sleep(1)
+                self.term_info[1] = "Select Space by entering 1-9"
+
+        self.redraw_user_term(term)
+
+        if self.update_board:
+            time.sleep(1)
+            self.end_of_turn(term)
+
+    def confirm_entry(self, term: blessed.Terminal) -> None:
+        """Confirm user pressed enter"""
+        if self.user_select_subgrid != 0 and self.current == 20:
+            self.term_info[1] = "Select Space by entering 1-9"
+        elif self.user_select_space != 0 and self.current == 30:
+            # show confirmation
+            self.term_info[0] = " "
+            self.term_info[1] = "Player selected:"
+            self.term_info[2] = (
+                f"Current: SubGrid {self.user_select_subgrid} "
+                f"| Space {self.user_select_space}"
+            )
+
+        self.redraw_user_term(term)
+
+    def end_of_turn(self, term: blessed.Terminal) -> None:
+        """End of Turn"""
+        self.current = 40
+
+        # update board
+        self.update_board = True
+
+        # circle back to top
+        self.wait_for_ready(term)
+
+    def confirm_good_move(self, board: Board) -> bool:
+        """Handle the entry of a bad move"""
+        # TODO need to finish this. maybe move to game logic?
+        # guilty until proven innocent
+        self.good_move = False
+
+        working_space_location = convert_to_space_location(self.user_select_space)
+        if (
+            board.collect_subgrid(str(self.user_select_subgrid - 1))[
+                working_space_location[0], working_space_location[1]
+            ]
+            == "·"

```

Comment:
Should this be a `not in ("X", "O")` to make it easier to change the nothing-here character?

Reply:
sure